Single tank and TTV tweaks

LorenLuke opened the pull request:
Changes the way single tank bombs (and even TTVs) now function, including their construction, a safety relief valve, having explosions scale with gas amount and the addition of shrapnel.

Thanks to Daranz for helping me make my first PR, and @Asanadas for getting the idea on paper in the first place.

LorenLuke committed:
Changes behavior, creation and dismantleing of single-tank devices. Also changes the leak/rupture/explosion logic for tanks in general.

Tanks now have emergency relief valves that open at over 173 C or 30kPA. They can be welded shut (don't move doing it though!)
Tanks generate shrapnel when exploding or rupturing. This includes TTVs.
Tank explosions scale logarithmically with amount of moles in the tank (TTVs do more damage than single tanks even at same pressure).
Single tank devices more difficult to construct: use wire, then igniter assembly. Use wirecutters to remove.
Gas in the tank is merged into the ambient atmosphere on exploding or rupturing.

LorenLuke committed:
Minor followup changes.

Added descriptors to tanks for tweaked info.
Created singletank bomb spawner (/obj/effect/bombspawner/onetankbomb).
Removed references to obj/item/devices/onetankbomb.

mwerezak commented on code/game/objects/items/weapons/tanks/tanks.dm:
```diff
@@ -86,15 +124,99 @@ var/list/global/tank_gauge_cache = list()
-
 	ui_interact(user)
 
+// There's GOT to be a better way to do this
```
By making interaction with the assembly happen through the GUI, most likely.

LorenLuke replied:
I dunno nanoUI, or how it works, so there's currently this. Plus assemblies are still broken if they're not in your hand, anyway.

mwerezak commented on code/game/objects/items/weapons/tanks/tanks.dm:
```diff
@@ -249,63 +371,297 @@ var/list/global/tank_gauge_cache = list()
+		src.overlays += "bomb_assembly"
+
+	if(src.proxyassembly.assembly)
+		var/icon/test = getFlatIcon(src.proxyassembly.assembly)
```
Are you sure this cannot be done with images instead?

LorenLuke replied:
Same code as the fueltanks. Besides, it makes it really visible if one's put down, which is good, methinks.

mwerezak replied:
I guess they will both have to be updated sometime.

LorenLuke committed:
Fixed a few issues with unclosed tags in strings.

Also making sure NOT to include baystation12.int this time...

mwerezak commented on code/game/objects/items/weapons/tanks/tanks.dm:
```diff
@@ -249,63 +371,297 @@ var/list/global/tank_gauge_cache = list()
+			var/env_pressure = environment.return_pressure()
+			var/tank_pressure = src.air_contents.return_pressure()
+
+			var/release_ratio = min(max( 0.02, sqrt((tank_pressure-env_pressure)/tank_pressure) ),1)
```
Clamp() would be nice for readability.

mwerezak commented on baystation12.int:
```diff
@@ -1,6 +1,9 @@
 // BEGIN_INTERNALS
 /*
-MAP_ICON_TYPE: 0
+DIR: code code\_onclick code\datums code\datums\trading code\game code\game\objects code\game\objects\effects code\game\objects\effects\spawners code\game\objects\items code\game\objects\items\weapons code\game\objects\items\weapons\tanks code\modules code\modules\admin code\modules\admin\secrets code\modules\admin\secrets\admin_secrets code\modules\assembly code\modules\mob code\modules\mob\livi ...
```
Committed .int

mwerezak commented on code/game/objects/items/weapons/tanks/tanks.dm:
```diff
@@ -249,63 +371,297 @@ var/list/global/tank_gauge_cache = list()
+
+			pressure = air_contents.return_pressure()
+			var/strength = ((pressure-TANK_FRAGMENT_PRESSURE)/TANK_FRAGMENT_SCALE)
+			var/mult = max(0.2,1 + log(2.6, (air_contents.total_moles/29)) ) //1 + log base 2 of standard sized full (room temperature) tanks
```
I'm still of the opinion that it would be better to scale off of volume instead of moles. Enthalpy is U + PV after all. I'm curious why you chose moles.

GinjaNinja32 replied:
`1013.25 x 70 = 70927.5`: yep
`70927.50/(8.32 x 293.15) = 32`: nope, it's 29.08047, you should check your calculator :P

The 'exact' value, ie ignoring game stuff and treating it as a perfect simulation, is 29.09984 mol; the difference between this and the above 29.08047 is that `R = 8.3144621 != 8.32`

Either way this should be an expression of constants; BYOND should be simplifying those down to a plain value at compilation, so it doesn't make a difference when executing, only when understanding; it's easier to understand what `10*ONE_ATMOSPHERE * 70 / (R_IDEAL_GAS_EQUATION * T20C)` (pV/RT) actually _means_ than a magic constant `29`.

Make it a define somewhere in code/__defines if you like, even better if you find a relevant define for max tank pressure rather than hardcoding 10 atmospheres.

Asanadas commented:
Glad to see this improved version is up!

mwerezak commented on code/game/objects/items/weapons/tanks/tanks.dm:
```diff
@@ -249,63 +371,297 @@ var/list/global/tank_gauge_cache = list()
+			var/list/target_turfs = getcircle(T, 7)
+			var/fragments_per_projectile = round(fragments/target_turfs.len)
+
+			for(var/turf/O in target_turfs)
```
This should be in a proc, bad copypasta.

LorenLuke commented on code/game/objects/items/weapons/tanks/tanks.dm:
```diff
@@ -249,63 +371,297 @@ var/list/global/tank_gauge_cache = list()
+
+			pressure = air_contents.return_pressure()
+			var/strength = ((pressure-TANK_FRAGMENT_PRESSURE)/TANK_FRAGMENT_SCALE)
+			var/mult = max(0.2,1 + log(2.6, (air_contents.total_moles/29)) ) //1 + log base 2 of standard sized full (room temperature) tanks
```
More in the idea that n -> V when T,P are constant. Therefore n1 @ PV1 = 2 \* n1 @ PV2, if V1 = 2\* V2.

So since it scales with volume (assuming all else is equal) I view it as easier to determine than V for the sake of calculation.

mwerezak replied:
Instead of using standard tank volume 70 L as our reference volume now you have to use reference moles. Plus I'm not even sure what 29 mol is calculated from, that really should be an expression of constants there.

LorenLuke replied:
~29.2 moles is 1013.3 kPA @ 20C, aka 1 full airtank.

MATH:
PV = nRT
1013.25 x 70 = 70927.5
70927.50/(8.32 x 293.15) =32

While the math seems to be off, fill a tank to 10 ATM and VV it. Numbers are molar, iirc.

mwerezak commented on html/changelogs/LorenLuke - Singletanks.yml:
```diff
@@ -0,0 +1,14 @@
+delete-after: True
+
+changes: 
+  - rscdel: "Removed bomb.dm from baystation12.dme."
```
Players don't know or care what source files you touched.

Nor do they care about what awful code you excised, unless it removed some beloved item or feature.

Describe how the overpressurized tanks will behave, don't just state that you changed some logic.

LorenLuke replied:
There's descriptions in tanks.dm. If people bother to read them and examine them, there won't be an issue.

mwerezak replied:
I'm not sure how that's an excuse to not clean this up.

LorenLuke replied:
It's not an excuse of any type. It's addressing your concern about players having no knowledge of the changes in terms of how over pressurized tanks are different. As for cleaning things up, I frankly don't care if players don't care. It's a changelog, it details the changes I made, and it's not unlike those changes aren't thoroughly detailed and available to be seen by examining a tank.

mwerezak replied:
This changelog is player facing, you shouldn't need to mention that you removed this or that file. It's clutter and pointless, we can see the commit log and PR diff.

It's an excuse because the descriptions in tank.dm are not relevant to my comment in any way, I can't imagine what else you could mean to say by bringing it up.

LorenLuke replied:
> Describe how the overpressurized tanks will behave, don't just state that you changed some logic.

`description_info = "These tanks are utilised to store any of the various types of gaseous substances on the station. They can be attached to various portable atmospheric devices on station to be filled or emptied. Each tank is fitted with an emergency relief valve. This relief valve will open if the tank is pressurised to over ~3000kPa or heated to over 173ºC. The valve itself will close after expending most or all of the contents into the air. Filling a tank such that experiences ~4000kPa of pressure will cause the tank to rupture, spilling out its contents and destroying the tank. Tanks filled over ~5000kPa will rupture rather violently, exploding with significant force."

description_antag = "Each tank may be incited to burn by attaching wires and an igniter assembly, though the igniter can only be used once and the mixture only burn if the igniter pushes a flammable gas mixture above the minimum burn temperature (126ºC).  Wired  and assembled tanks may be disarmed with a set of wirecutters. Any exploding or rupturing tank will generate shrapnel, assuming their relief valves have been welded beforehand. Even if not, they can be incited to expel hot gas on ignition if pushed above 173ºC. Relatively easy to make, the single tank bomb requries no tank transfer valve, and is still a fairly formidable weapon that can be manufactured from any tank."`

mwerezak replied:
I'm not sure what you're trying to say. Do you mean you intend to copy that into the changelog? 

The values you have given to `description_info` and `description_antag` are completely irrelevant to the changelog and you have not said anything to indicate what logical connection your response has to my comment.

It's pretty obvious you don't understand something here. I'm not sure how else to help you since you won't explain what it is. 

The best I can do is just repeat what I've said already: clean up the irrelevant clutter out of this changelog file. I've made it pretty clear how to tell what is relevant to a changelog and what isn't

LorenLuke replied:
Oh! I thought you were saying 'nowhere were the specific changes in behavior documented, and therefore, shame', not 'nowhere were the specific changes in behavior documented _in the changelog_, and therefore, shame'. I'll pop these into the changes next commit.

mwerezak commented on code/game/objects/effects/spawners/bombspawner.dm:
```diff
@@ -190,3 +192,62 @@
+
+
+	K.overlays += "bomb_assembly"
+	var/icon/test = getFlatIcon(K.proxyassembly.assembly)
```
Call update_gauge() or whatever don't copypaste this here.

LorenLuke replied:
Except update_gauge() won't work on phoron tanks, because they have no gauge.

mwerezak replied:
Then fix that. The overlay isn't part of the gauge, why is it in update_gauge()? Move it somewhere more appropriate.

LorenLuke replied:
In update_gauge, it cuts all overlays and adds the one for the gauge, thus any pressure changes would get rid of any other overlays. I'll sort it out via another procedure.

mwerezak replied:
Sure, otherwise I'll see if I can sort it out sometime.

mwerezak commented on code/game/objects/items/weapons/tanks/tanks.dm:
```diff
@@ -249,63 +371,297 @@ var/list/global/tank_gauge_cache = list()
+				)
+
+
+			T.hotspot_expose(700, 50, 1)
```
I know you're just reusing the existing code but it would be nice if you could use the gas mix values here instead of these constants. Look up the definition of hotspot_expose()

mwerezak commented on code/game/objects/items/weapons/tanks/tanks.dm:
```diff
@@ -86,15 +124,99 @@ var/list/global/tank_gauge_cache = list()
+			user << "<span class='notice'>You need to wire the device up first.</span>"
+
+
+	if((istype(W, /obj/item/weapon/weldingtool) && W:welding))
```
This is _not_ the correct way to tell if a welding tool is on.

LorenLuke replied:
What would be?

mwerezak replied:
If you look at examples throughout the code, when someone uses a welder you need to call `W.remove_fuel(0,user)`, that will return 1 if the welder can actually be used in addition to checking if they are wearing proper protection and applying eye damage, etc.

mwerezak commented:
I like the idea, the issues need to be dealt with though. I'll go over it more tomorrow when I'm at an actual computer.

LorenLuke committed:
Fixed a lot of mwerezak/HarpyEagle's concerns.

LorenLuke commented:
Paging @mwerezak for inspection.

LorenLuke committed:
Put fragmentation code into its own proc. Did a minor tweak to a check and effect about leaking tanks.

LorenLuke committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into Single-tank-tweaks

LorenLuke committed:
Final commit. Removing reference to /obj/item/device/onetankbomb. Changes W-class of tanks to be smaller (as one-tankers make them smaller, and TTVs are smaller yet).

LorenLuke commented:
@Mloc , @mwerezak 

Final inspection.

LorenLuke commented:
@Mloc , @mwerezak

Final inspection.

mwerezak commented on code/game/objects/items/weapons/tanks/tanks.dm:
```diff
@@ -14,24 +14,51 @@ var/list/global/tank_gauge_cache = list()
 
 	flags = CONDUCT
 	slot_flags = SLOT_BACK
-	w_class = 4
```
Why?

LorenLuke replied:
> (as one-tankers make [the tanks] smaller, and TTVs are smaller yet)

Makes the tanks similarly sized to the devices they're designed to replace.

mwerezak replied:
So? How does that justify making tanks smaller?

LorenLuke replied:
Look, I'm not going to argue it out. I've given my reasoning. If you disagree with it, you have full reign to add changes to change it back. I'm gone for eight days and myself cannot. Truth be told with a 70 liter backpack looking like [http://www.sportsmanswarehouse.com/img/products/original/jansport_katahdin_70_multiday_pack_1395932_1.jpg](url), they shouldn't be able to fit at all. Do what you think is best.

mwerezak commented on code/game/objects/items/weapons/tanks/tanks.dm:
```diff
@@ -245,68 +412,232 @@ var/list/global/tank_gauge_cache = list()
-		if(integrity <= 0)
 			var/turf/simulated/T = get_turf(src)
+			T.hotspot_expose(src.air_contents.temperature, 70, 1)
 			if(!T)
```
You reference T immediately before checking if it actually has a value?

LorenLuke replied:
So I did. Agree that it should be after as an else.

mwerezak commented:
Some other things too probably. Easier for me just to PR changes.

mwerezak commented on code/game/objects/items/weapons/tanks/tanks.dm:
```diff
@@ -245,68 +412,232 @@ var/list/global/tank_gauge_cache = list()
+				round(min(BOMBCAP_DVSTN_RADIUS, ((mult)*strength)*0.20)),
+				round(min(BOMBCAP_HEAVY_RADIUS, ((mult)*strength)*0.40)),
+				round(min(BOMBCAP_LIGHT_RADIUS, ((mult)*strength)*0.85)),
+				round(min(BOMBCAP_FLASH_RADIUS, ((mult)*strength)*1.20)),
```
Why did you change all of these?

LorenLuke replied:
Bomb cap preserved, while the scalar, multiplier and sizing allows for the scaling of the bombs size to be tied to moles (which corresponds directly to volume as stated in prior conversation), these scalings roughly preserve TTVs damage while single tank bombs are less powerful and significantly less likely to breach.

LorenLuke commented:
@mwerezak If you want to PR any changes, go ahead. I'm away from my computer until a week from Wednesday, and the Lion's share of the work is done. Feel free to tweak it as you see fit; test various values first, but barring any code optimizations, I'm willing to ship it out.

mwerezak commented:
Someday I will get to this, already have done a bunch of work on it but not yet complete.

LorenLuke committed:
Fixes w_class on TTVs being small (now w_class = 4 when tanks are attached) even with tanks attached (so that you can have more than double the number of tanks in a bag when they're attached to TTVs), as well as increasing the w_class of bombs to tank defaults (w_class = 4).
Added provisions for compatibility with trader files and obj/item/device/onetankbomb. (Merge conflicts can ignore changes on this PR to trader files).

LorenLuke commented:
@Mloc :

In the wake of about two months of waiting, plus this [thread](url=https://baystation12.net/forums/threads/bye.2576/), I don't imagine that this will be looked at by them anytime soon. Requesting DNM be removed and PR be merged, as the latest commit addresses one of their concerns, plus a few more of my own.

Conflicts in the 'code/datums/trading' folder can overwrite the files in this PR, as a onetankbomb bomb spawner was added to allow compatibility between this PR and the current merchant code.

Kasuobes commented:
Just unstage the two files then.

LorenLuke commented:
@Kasuobes You say that like I know what I'm doing and totally not using a Tortoise Git to flounder about.

I'll attempt to without utterly fouling up everything in the PR, but if the experts can manage to resolve conflicts that way, that can work as well.

LorenLuke commented:
Gah, shit... this is why I leave things to the experts... hang on...

How the fuck do I reopen this?

LorenLuke commented:
Yeah, no... after this lovely display of my incompetence, they know how to fix any merge problems with the above instructions. New opening, final revision.

Kasuobes commented:
Go into both files and manually merge your changes, then commit again with the changed files.

LorenLuke committed:
Fixes w_class on TTVs being small (now w_class = 4 when tanks are attached) even with tanks attached (so that you can have more than double the number of tanks in a bag when they're attached to TTVs), as well as increasing the w_class of bombs to tank defaults (w_class = 4). Allows for obj/item/device/ontankbomb references. Fixes Merge commits.

LorenLuke commented:
Well, I attempted to manually merge the changes, and it got one of the files fixed, but for some reason copy/pasting from github doesn't fix one of the merge conflicts. I'm sure this can be improved upon, but I'm not sure how, at this time.

Kasuobes commented:
https://help.github.com/articles/resolving-a-merge-conflict-from-the-command-line/
